---
title: '[JavaScript] ClassMethod を代入すると意図しない動きをする'
excerpt: 'ClassFunction を変数に代入すると、参照が代入されるだけかと思ったが意図しない動きをするので原因と対策をメモしておく'
coverImage: 
  url: 'https://i.imgur.com/Fl0wbEb'
  aspectRatio: '4/3'
date: '2022-03-19T20:00:00+09:00'
ogImage:
  url: 'https://i.imgur.com/Fl0wbEb'
  aspectRatio: '4/3'
tags: ['JavaScript']
---


## 概要
* ClassFunction を変数に代入すると、参照が代入されるだけかと思ったが意図しない動きをする
* 代入する際に this が暗黙的に変更されるのが原因
* function を代入する際に代入する際にbindでthisを明示的に束縛してあげれば解決する

## 意図しない動きをするパターン

```ts
class SomeClass {
  param: string;
  
  constructor() {
    this.param = "test";
  }
  
  say() {
    console.log(this)
    console.log(`${this.param}`);
  }
}

  

const test = new SomeClass()
test.say()
  

const testMethod = test.say
testMethod()
```
[playground](https://www.typescriptlang.org/play?ssl=18&ssc=13&pln=1&pc=1#code/MYGwhgzhAEDKD2BbApgYXFaBvAUNf0ADmAE5iIBc0EALiQJYB2A5gNw54HDyO0kCuwGvBIAKAJTZOBAjQAW9CADpiZRNAC80AEQ1ktbexkBfDjIhgAnhKkyZ3XvBDIlIeM1HzF46XYcQnFzcPAAMAEiwvZVVyYxDxIwJTUxx-Gmg9Wk1oRmQAdzgkNAwICRxMmiULax9UniyKgFlkeXgAE2yKqqty-Rpm1raJIA) 

このまま実行すると、最終行で `console.log(this)` が呼ばれた際に this が undefined になりその次の行でエラーが出る

## 解決策

```ts
class SomeClass {
  param: string;
  
  constructor() {
    this.param = "test";
  }
  
  say() {
    console.log(this)
    console.log(`${this.param}`);
  }
}

  

const test = new SomeClass()
test.say()
  

const testMethod = test.say.bind(test)
testMethod()
```
[playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDKD2BbApgYXFaBvAUNf0ADmAE5iIBc0EALiQJYB2A5gNw54HDyO0kCuwGvBIAKAJTZOBAjQAW9CADpiZRNAC80AEQ1ktbexkBfDjIhgAnhKkyZ3XvBDIlIeM1HzF46XYcQnFzcPAAMAEiwvZVVyYxDxIwJTUxx-Gmg9Wk1oRmQAdzgkNAwICRxMmiULax9UniyKgFlkeXgAE2yKqqslACMmNs99Gh8mlrl2iSA) 


`bind` で this を明示的に束縛すると意図した通りの動きをする